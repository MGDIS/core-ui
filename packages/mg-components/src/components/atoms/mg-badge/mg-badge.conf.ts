import type { Variant } from '../../../types';

/**
 * List of all possibles variants
 */
export const variants = ['info', 'primary', 'secondary', 'success', 'warning', 'danger', 'text-color'] as const;

/**
 * Variant type from variants
 */
export type BadgeVariantType = Variant & (typeof variants)[number];
