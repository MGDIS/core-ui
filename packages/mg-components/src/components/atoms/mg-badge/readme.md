## Design

The badge is always placed on top of, or next to the element it is for.

The badge displays a number (can be followed by the `+` character) or a punctuation character.

## Specs

![](./doc/img/mg-badge-specs.png)

## Theming

![](./doc/img/mg-badge-styles.png)

<!-- Auto Generated Below -->


## Properties

| Property             | Attribute | Description                                                        | Type                                                                                       | Default       |
| -------------------- | --------- | ------------------------------------------------------------------ | ------------------------------------------------------------------------------------------ | ------------- |
| `label` _(required)_ | `label`   | Badge label. Include short description. Required for accessibility | `string`                                                                                   | `undefined`   |
| `outline`            | `outline` | Define if button is using outline style                            | `boolean`                                                                                  | `undefined`   |
| `value` _(required)_ | `value`   | Badge value                                                        | `number \| string`                                                                         | `undefined`   |
| `variant`            | `variant` | Define badge variant                                               | `"danger" \| "info" \| "primary" \| "secondary" \| "success" \| "text-color" \| "warning"` | `variants[0]` |


## CSS Custom Properties

| Name                      | Description                                                                                                                  |
| ------------------------- | ---------------------------------------------------------------------------------------------------------------------------- |
| `--mg-c-badge-font-size`  | Defines badge font size. Default value: `1.1rem`.                                                                            |
| `--mg-c-badge-size`       | Defines badge height and min-width. Default value: `--mg-b-size-2w`.                                                         |
| `--mg-c-badge-text-color` | Defines badge font color when used with the `text-color` variant and without `outline`. Default value: `--mg-b-color-light`. |


## Dependencies

### Used by

 - [mg-action-more](../../molecules/mg-action-more)
 - [mg-menu-item](../../molecules/menu/mg-menu-item)
 - [mg-tabs](../../molecules/mg-tabs)

### Graph
```mermaid
graph TD;
  mg-action-more --> mg-badge
  mg-menu-item --> mg-badge
  mg-tabs --> mg-badge
  style mg-badge fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
