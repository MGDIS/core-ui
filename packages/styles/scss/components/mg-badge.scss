@use 'sass:list';
@use '../variables' as variables;
@import '../utils/a11y';

/*!
 * @prop --mg-c-badge-size: Defines the height and minimum width of the badge. The default value is `--mg-b-size-16`.
 * @prop --mg-c-badge-font-size: Defines the font size of the badge. The default value is `--mg-b-font-size-small`.
 * @prop --mg-c-badge-text-color: Defines the font color of the badge when used with the `text-color` variant and without `outline`. The default value is `--mg-b-color-light`.
 */

/* Styles */
.mg-c-badge {
  min-width: var(--mg-c-badge-size);
  height: var(--mg-c-badge-size);
  box-sizing: border-box;
  padding: 0 var(--mg-b-size-4);
  border: var(--mg-b-size-border) solid;
  border-radius: calc(var(--mg-c-badge-size) / 2);
  font-size: var(--mg-c-badge-font-size);
  font-weight: 600;
  line-height: var(--mg-c-badge-size);
  text-align: center;

  // Variant
  @each $name, $palette, $index in list.join(list.join(variables.$brands, variables.$semantics), 'text-color') {
    &.mg-c-badge--#{$name} {
      @if $name == 'text-color' {
        border-color: currentcolor;
        background-color: currentcolor;

        .mg-c-badge__value {
          color: var(--mg-c-badge-text-color, var(--mg-b-color-light));
        }
      } @else {
        border-color: var(--mg-b-color-#{$name});
        background-color: var(--mg-b-color-#{$name});
        color: var(--mg-b-color-text-on-#{$name});
      }

      &.mg-c-badge--outline {
        background-color: transparent;

        @if $name == 'text-color' {
          .mg-c-badge__value {
            color: currentcolor;
          }
        } @else {
          color: var(--mg-b-color-#{$name});
        }
      }
    }
  }
}

.mg-c-badge,
mg-badge {
  display: inline-flex;
  align-items: center;
  justify-content: center;
}
