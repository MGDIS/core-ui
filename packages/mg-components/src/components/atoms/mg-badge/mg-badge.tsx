import { Component, h, Prop, State, Watch } from '@stencil/core';
import { variants, BadgeVariantType } from './mg-badge.conf';
import { ClassList } from '../../../utils/components.utils';
@Component({
  tag: 'mg-badge',
  styleUrl: '../../../../node_modules/@mgdis/styles/dist/components/mg-badge.css',
  shadow: true,
})
export class MgBadge {
  /************
   * Internal *
   ************/

  // Classes
  private readonly classOutline = `mg-c-badge--outline`;

  /**************
   * Decorators *
   **************/

  /**
   * Badge value
   */
  @Prop() value!: string | number;
  @Watch('value')
  validateValue(newValue: MgBadge['value']): void {
    if (`${newValue}`.match(/^(\d+\+*|[?*!a-z])$/i) === null) {
      throw new Error('<mg-badge> prop "value" must be integer or ponctuation character.');
    }
  }

  /**
   * Badge label. Include short description.
   * Required for accessibility
   */
  @Prop() label!: string;
  @Watch('label')
  validateLabel(newValue: MgBadge['label']): void {
    if (typeof newValue !== 'string' || newValue.trim() === '') {
      throw new Error('<mg-badge> prop "label" is required.');
    }
  }

  /**
   * Define button variant
   */
  @Prop() variant?: BadgeVariantType = variants[0]; // info
  @Watch('variant')
  validateVariant(newValue: MgBadge['variant'], oldValue?: MgBadge['variant']): void {
    if (!variants.includes(newValue)) {
      throw new Error(`<mg-badge> prop "variant" must be one of: ${variants.join(', ')}.`);
    } else {
      if (oldValue !== undefined) {
        this.classCollection.delete(`mg-c-badge--${oldValue}`);
      }
      this.classCollection.add(`mg-c-badge--${newValue}`);
    }
  }

  /**
   * Define if button is using outline style
   */
  @Prop() outline: boolean;
  @Watch('outline')
  validateOutline(newValue: MgBadge['outline']): void {
    if (newValue) this.classCollection.add(this.classOutline);
    else this.classCollection.delete(this.classOutline);
  }

  /**
   * Component classes
   */
  @State() classCollection: ClassList = new ClassList(['mg-c-badge']);

  /**
   * Check if props are well configured on init
   */
  componentWillLoad(): void {
    this.validateValue(this.value);
    this.validateLabel(this.label);
    this.validateVariant(this.variant);
    this.validateOutline(this.outline);
  }

  /**
   * Render
   * @returns HTML Element
   */
  render(): HTMLElement {
    return (
      <span class={this.classCollection.join()}>
        <span class="mg-c-badge__value">{this.value}</span>
        <span class="mg-u-visually-hidden">{this.label}</span>
      </span>
    );
  }
}
