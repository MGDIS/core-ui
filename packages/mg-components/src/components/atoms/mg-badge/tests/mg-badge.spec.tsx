import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgBadge } from '../mg-badge';
import { variants } from '../mg-badge.conf';

const getPage = args =>
  newSpecPage({
    components: [MgBadge],
    template: () => <mg-badge {...args}></mg-badge>,
  });

describe('mg-badge', () => {
  describe.each(variants)('variant %s', variant => {
    test.each([true, false])('outiline %s', async outline => {
      const { root } = await getPage({ variant, outline, value: 1, label: 'Batman' });
      expect(root).toMatchSnapshot();
    });
  });

  test.each([1, 100, '!', '?', '99+', '+', '99?'])('value %s', async value => {
    const { root } = await getPage({ value, label: 'Batman' });
    expect(root).toMatchSnapshot();
  });

  test('Should replace classes on variant/ouline changes', async () => {
    const page = await getPage({ value: 1, label: 'Blu' });
    const element = page.doc.querySelector('mg-badge');
    let classInfo = element.shadowRoot.querySelector('.mg-c-badge--info');
    let classOutline = element.shadowRoot.querySelector('.mg-c-badge--outline');

    expect(classInfo).not.toBeNull();
    expect(classOutline).toBeNull();

    // Change variant
    element.variant = 'danger';
    await page.waitForChanges();

    classInfo = element.shadowRoot.querySelector('.mg-c-badge--info');
    const classDanger = element.shadowRoot.querySelector('.mg-c-badge--danger');

    expect(classInfo).toBeNull();
    expect(classDanger).not.toBeNull();

    // Change outline
    element.outline = true;
    await page.waitForChanges();

    classOutline = element.shadowRoot.querySelector('.mg-c-badge--outline');
    expect(classOutline).not.toBeNull();
  });

  describe('errors', () => {
    test.each(['', ' ', undefined])('Should throw error, case invalid label prop', async label => {
      expect.assertions(1);
      try {
        await getPage({ value: 1, label });
      } catch (err) {
        expect(err.message).toEqual(`<mg-badge> prop "label" is required and must be a string. Passed value: ${label}.`);
      }
    });

    test.each(['', 'Batman'])('Should throw error, case invalid variant prop', async variant => {
      expect.assertions(1);
      try {
        await getPage({ variant, value: 1, label: 'Batman' });
      } catch (err) {
        expect(err.message).toEqual(`<mg-badge> prop "variant" must be one of: ${variants.join(', ')}. Passed value: ${variant}.`);
      }
    });

    test.each(['', 'Batman', 'BATMAN', 'batman', 'Batman+', 'Batman!'])('Should throw error, case invalid value prop %s', async value => {
      expect.assertions(1);
      try {
        await getPage({ value, label: 'Batman' });
      } catch (err) {
        expect(err.message).toEqual(`<mg-badge> prop "value" is required and must be integer and/or special character. Passed value: ${value}.`);
      }
    });
  });
});
