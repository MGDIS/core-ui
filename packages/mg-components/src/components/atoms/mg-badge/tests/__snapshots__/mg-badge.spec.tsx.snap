// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-badge value ! 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--info">
      <span class="mg-c-badge__value">
        !
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge value + 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--info">
      <span class="mg-c-badge__value">
        +
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge value ? 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--info">
      <span class="mg-c-badge__value">
        ?
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge value 1 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--info">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge value 99+ 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--info">
      <span class="mg-c-badge__value">
        99+
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge value 99? 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--info">
      <span class="mg-c-badge__value">
        99?
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge value 100 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--info">
      <span class="mg-c-badge__value">
        100
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant danger outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--danger">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant danger outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--danger mg-c-badge--outline">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant info outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--info">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant info outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--info mg-c-badge--outline">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant primary outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--primary">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant primary outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--outline mg-c-badge--primary">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant secondary outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--secondary">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant secondary outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--outline mg-c-badge--secondary">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant success outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--success">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant success outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--outline mg-c-badge--success">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant text-color outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--text-color">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant text-color outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--outline mg-c-badge--text-color">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant warning outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--warning">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;

exports[`mg-badge variant warning outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-c-badge mg-c-badge--outline mg-c-badge--warning">
      <span class="mg-c-badge__value">
        1
      </span>
      <span class="mg-u-visually-hidden">
        Batman
      </span>
    </span>
  </mock:shadow-root>
</mg-badge>
`;
